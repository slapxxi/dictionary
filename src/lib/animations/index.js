// @flow
import { chain, tween, easing, stagger } from 'popmotion';
import { percent } from 'style-value-types';

type Styler = Object;

type Dimension = 'x' | 'y';

type Config = {
  styler: Styler,
  to?: number,
  ease?: any,
  delay?: number,
  duration?: number,
  dimension?: Dimension,
};

function slideIn({
  styler,
  to = -100,
  delay = 100,
  dimension = 'x',
}: Config) {
  return stagger(
    [
      translate({ styler, dimension, to }),
      opacity({ styler, to: 1 }),
    ],
    delay,
  ).pipe(([translate, op = { opacity: styler.get('opacity') }]) => ({
    translate: translate[dimension],
    opacity: op.opacity,
  }));
}

function slideOut({
  styler,
  to = -100,
  delay = 100,
  dimension = 'x',
}: Config) {
  return stagger(
    [
      opacity({ styler, to: 0 }),
      translate({ styler, dimension, to }),
    ],
    delay,
  ).pipe(
    ([
      { opacity },
      dim = { [dimension]: percent.parse(styler.get(dimension)) },
    ]) => ({
      opacity,
      translate: dim[dimension],
    }),
  );
}

function pop({ styler, duration = 100, to = 2 }: Config) {
  return chain(
    tween({ from: styler.get('scale'), to: { scale: to }, duration }),
    tween({ from: { scale: to }, to: { scale: 1 } }),
  );
}

function translate({
  styler,
  to,
  dimension = 'x',
  ease = easing.linear,
}: Config) {
  return tween({
    from: { [dimension]: percent.parse(styler.get(dimension)) },
    to: { [dimension]: to },
    duration: 300,
    easing: ease,
  });
}

function opacity({ styler, to, ease = easing.backIn }: Config) {
  return tween({
    from: {
      opacity: styler.get('opacity'),
    },
    to: { opacity: to },
    easing: ease,
  });
}

export { slideIn, slideOut, pop, translate, opacity };
