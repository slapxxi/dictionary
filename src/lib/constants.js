// @flow
const colors = {
  red: 'tomato',
  blue: '#17F',
  white: 'white',
  black: 'black',
};

export { colors };
