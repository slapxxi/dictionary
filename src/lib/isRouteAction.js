// @flow
import { NOT_FOUND } from 'redux-first-router';

function isRouteAction(action) {
  if (action.type === NOT_FOUND) {
    return true;
  }
  return action.type.toLowerCase().search('_route') !== -1;
}

export default isRouteAction;
