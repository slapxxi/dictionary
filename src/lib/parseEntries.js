// @flow
import _ from 'lodash';
import type { DictionaryEntry } from '../store/types';

type Params = DictionaryEntry;

function parseEntries(
  entries: Array<string>,
  params?: Params,
): Array<DictionaryEntry> {
  return entries.map((entry, index) => {
    const data = entry.split(';');
    return {
      id: index,
      word: data[0].trim(),
      transcription: data[1].trim(),
      definition: data[2].trim(),
      examples: data[3]
        .trim()
        .split(':')
        .map((i) => i.trim())
        .map(_.capitalize),
      thesaurus: data[4].trim().split(','),
      learnt: false,
      viewCount: 0,
      ...(params || {}),
    };
  });
}

export default parseEntries;
