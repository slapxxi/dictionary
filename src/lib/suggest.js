// @flow
type Query = string;

type Predicate<T> = (T) => string;

function suggest<T>(
  query: Query,
  data: Array<T>,
  predicate: Predicate<T>,
): Array<T> {
  const q = query.trim();
  if (q === '') {
    return [];
  }
  return data.filter((item) =>
    new RegExp(`${q}`, 'i').test(predicate(item)),
  );
}

export default suggest;
