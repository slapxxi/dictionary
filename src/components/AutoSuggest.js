// @flow
import { isEmpty } from 'lodash';
import React, { Component } from 'react';
import glamorous from 'glamorous';
import { suggest } from '../lib';

type Props = { data: Array<Datum>, query: any };

type Datum = { id: number, text: string };

class AutoSuggest extends Component<Props> {
  render() {
    const { data, query } = this.props;
    if (!data || isEmpty(data) || !query) {
      return null;
    }
    const matches = suggest(query, data);
    return (
      <Container className="autosuggest">
        <ul>
          {matches.map((m) => (
            <ListItem key={m.id}>{m.text}</ListItem>
          ))}
        </ul>
      </Container>
    );
  }
}

const Container = glamorous.div({
  display: 'flex',
  flex: 1,
});

const ListItem = glamorous.li({
  padding: '12px 0',
  textDecoration: 'underline',
  textAlign: 'center',
});

export default AutoSuggest;
