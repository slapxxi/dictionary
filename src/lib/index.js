// @flow
import isRouteAction from './isRouteAction';
import suggest from './suggest';
import parseEntries from './parseEntries';

export { isRouteAction, suggest, parseEntries };
