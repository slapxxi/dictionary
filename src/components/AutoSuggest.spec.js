import React from 'react';
import { render } from 'enzyme';
import { suggest } from '../lib';
import AutoSuggest from './AutoSuggest';

const data = [{ word: 'test' }];

jest.mock('../lib', () => ({
  suggest: jest
    .fn()
    .mockImplementation(() => [
      { id: 0, word: 'abide' },
      { id: 1, word: 'abysmal' },
    ]),
}));

beforeEach(() => {
  suggest.mockClear();
});

it('renders suggestions', () => {
  const component = render(<AutoSuggest data={data} query="test" />);
  expect(component).toMatchSnapshot();
});

it('invokes suggest correctly', () => {
  render(<AutoSuggest data={[{ word: 'test' }]} query="test" />);
  expect(suggest.mock.calls[0][0]).toEqual('test');
  expect(suggest.mock.calls[0][1]).toEqual([{ word: 'test' }]);
  expect(suggest.mock.calls[0][2]).toBeDefined();
});

it('renders nothing if data is empty', () => {
  const component = render(<AutoSuggest data={[]} query="test" />);
  expect(component).toMatchSnapshot();
});

it('renders nothing if data not provided', () => {
  const component = render(<AutoSuggest query="query" />);
  expect(component).toMatchSnapshot();
});

it('renders nothing if query not provided', () => {
  const component = render(<AutoSuggest data={[]} />);
  expect(component).toMatchSnapshot();
});
