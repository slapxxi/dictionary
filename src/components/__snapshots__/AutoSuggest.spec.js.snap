// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders nothing if data is empty 1`] = `null`;

exports[`renders nothing if data not provided 1`] = `null`;

exports[`renders nothing if query not provided 1`] = `null`;

exports[`renders suggestions 1`] = `
.glamor-2,
[data-glamor-2] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex: 1;
  -webkit-flex: 1;
}

.glamor-0,
[data-glamor-0] {
  padding: 12px 0;
  text-align: center;
  font-size: 24px;
}

<div
  class="glamor-2 autosuggest"
>
  <ul>
    <li
      class="glamor-0"
    >
      abide
    </li>
    <li
      class="glamor-0"
    >
      abysmal
    </li>
  </ul>
</div>
`;
